### ISSUE TITLE
protobuf-kotlin dsl - Support <fieldName>OrNull getters for fields with explicit presence

### ISSUE THREAD

**Issue by berksean** [protocolbuffers/protobuf#16015]
**What language does this apply to?**
Generated code API for Kotlin

**Describe the problem you are trying to solve.**

For fields with presence, users often want to distinguish between a field that is present and set to the default value, and fields that are not present. The standard idiomatic way to distinguish is to use the hasX property/methods to check for presence before using the property to access the data.

Kotlin has null aware operators, and the primary IDE (Intellij) provides very helpful warnings/errors when there is potential to use a nullable value without checking for null. In this environment, it seems like providing accessors that will return `null` when a field is not present, and the value otherwise, would allow more natural / idiomatic kotlin to be written.

As the simplest example, instead of:

`val someValue = if (message.hasField) message.field else <some default>`

you can instead do:

`val someValue = message.fieldOrNull() ?: <some default>`

Of course this example is fairly trivial, but you can imagine with various chained null aware behaviors/nested field accesses, things could become much more convoluted if `if(...hasField)` checks needed to be placed at various levels in the expression.

**Describe the solution you'd like**
Modify the kotlin code generation in protoc to emit additional field accessors/properties for [explicit](https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md#presence-in-proto2-apis)  [presence](https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md#presence-in-proto3-apis) fields. These will return null if a field is not set, and the value otherwise.

e.g. `fun <field>OrNull() = if (has<field>) field else null`

No other changes are necessary, as the existing API can remain exactly the same.

One alternative that is a step further and somewhat controversial would be to make the standard accessors behave as described above for explicit presence fields. However, that seems like it could be quite confusing, and might necessitate adding additional properties like `<field>Default` to return the actual default value. On the whole, not sure I'd want it to go that far.

**Describe alternatives you've considered**
- Writing an additional `protoc` plugin to generate these accessors - Lots of the boilerplate code generation logic is already present in the java plugin which would need to be duplicated. Plus this idea seems like a generally useful addition that will improve interacting with protos in kotlin.
- Manually writing companion functions on an as-needed basis -- quite tedious
- ???

**Additional context**
N/A

**Comment by be-hase** [comment]
I also raised a similar issue before:
https://github.com/protocolbuffers/protobuf/issues/12935

As a result, I implemented my own protoc plugin:
https://github.com/be-hase/protoc-gen-kotlin-ext

**Comment by berksean** [comment]
Thanks for this @be-hase . Looks perfect. Would love these things to be native to the built in generated code, these seems like the next best thing.

I'll leave this issue around in case the protobuf team feels like making it a general addition.

**Comment by be-hase** [comment]
Yes, I have long hoped for it to be officially implemented. I would be very happy if protobuf team could consider it.
